lib/ts/authUtils.ts

```diff
@@ -0,0 +1,280 @@
+        req,
+        res,
+        tenantId,
+        userContext,
+    }: {
+        responseUser: User;
+        recipeUserId: RecipeUserId;
+        tenantId: string;
+        factorId: string;
+        isSignUp: boolean;
+        session?: SessionContainerInterface;
+        userContext: UserContext;
+        req: BaseRequest;
+        res: BaseResponse;
+    }): Promise<{ status: "OK"; session: SessionContainerInterface; user: User } | { status: "SIGN_IN_NOT_ALLOWED" }> {
+        const mfaInstance = MultiFactorAuthRecipe.getInstance();
+        const accountLinkingInstance = AccountLinking.getInstance();
+
+        // We check if sign in is allowed
+        if (
+            !isSignUp &&
+            !(await accountLinkingInstance.isSignInAllowed({ user: responseUser, tenantId, session, userContext }))
+        ) {
+            return { status: "SIGN_IN_NOT_ALLOWED" };
+        }
+
+        // Then run linking if necessary/possible.
+        // Note, that we are not re-using any information queried before the user signed in
+        // This functions calls shouldDoAutomaticAccountLinking again (we check if the app wants to link to the session user or not),
+        // which might return a different result. We could throw in this case, but it is an app code issue that that is fairly unlikely to go wrong.
+        // It should not happen in general, but it is possible that we end up with an unlinked user after this even though originally we considered this
+        // an MFA sign in.
+        // There are a couple of race-conditions associated with this functions, but it handles retries internally.
+        const postLinkUser = await accountLinkingInstance.createPrimaryUserIdOrLinkAccounts({
+            tenantId,
+            user: responseUser,
+            session,
+            userContext,
+        });
+
+        let respSession = session;
+        if (session !== undefined) {
+            const postLinkUserLinkedToSessionUser = postLinkUser.loginMethods.some(
+                (lm) => lm.recipeUserId === session.getRecipeUserId()
+            );
+            if (postLinkUserLinkedToSessionUser) {
+                // If the session user gets linked to the authenticating user we may end up in a situation where the userId of the session needs to change
+                if (postLinkUser.id !== session.getUserId()) {
+                    logDebugMessage(
+                        "postAuthChecks the session user id doesn't match the primary user id, so we are revoking all sessions and creating a new one"
+                    );
+                    await Session.revokeAllSessionsForUser(
+                        session.getUserId(),
+                        false,
+                        session.getTenantId(),
+                        userContext
+                    );
+
+                    // create a new session keeping the payloads from the input session
+                    respSession = await Session.createNewSession(
```

Comment:
I'm not really comfortable with doing this.. it feels weird. What do you think about these instead:
- If we reach a case where we have to link the session user to the secondary factor user, we instead return a support code; OR
- We just logout the user from their current session and so they have to sign in again.